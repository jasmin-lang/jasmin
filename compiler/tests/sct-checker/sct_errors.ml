open Jasmin
open Common

let () =
  let p = load_file "error_messages.jazz" in
  let check_fails f =
    match Sct_checker_forward.ty_prog Arch.is_ct_sopn p [ f ] with
    | exception Annot.AnnotationError (_loc, msg) ->
        Format.printf "Annotation error in %s: %t@." f msg
    | exception Utils.HiError e ->
        Format.printf "Failed as expected %s: %a@." f Utils.pp_hierror
          { e with err_loc = Lnone }
    | _ -> assert false
  in
  List.iter check_fails
    [
      "assign_msf";
      "syscall";
      "update_msf_not_trans";
      "update_msf_wrong_expr";
      "update_msf_not_msf";
      "msf_trans";
      "not_known_as_msf";
      "bad_poly_annot";
      "msf_in_export";
      "should_be_a_msf";
      "at_least_transient";
      "unbound_level";
      "inconsistent_constraint";
      "bad_modmsf";
      "bad_nomodmsf";
      "call_bad_nomodmsf";
      "call_bad_nomodmsf2";
      "call_bad_nomodmsf3";
      "call_modmsf_destroys";
      "ret_high";
      "ret_transient";
      "ret_msf";
      "public_arg";
      "unknown function";
      "need_declassify";
    ]
