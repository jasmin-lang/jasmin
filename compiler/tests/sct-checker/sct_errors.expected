Failed as expected assign_msf: speculative constant type checker: assignment to MSF variable x not allowed
Failed as expected syscall: speculative constant type checker: syscalls destroy msf variables, {
                             x } are required
Failed as expected update_msf_not_trans: speculative constant type checker: MSF is not Trans
Failed as expected update_msf_wrong_expr: speculative constant type checker: the expression false need to be equal to
                                          (x ==64u ((64u) 0))
Failed as expected update_msf_not_msf: speculative constant type checker: 
                                       the variables { x } need to be MSFs.
Failed as expected msf_trans: speculative constant type checker: MSF is Trans
                              (x ==64u ((64u) 0))
Failed as expected not_known_as_msf: speculative constant type checker: 
                                     the variables { x } need to be MSFs.
Failed as expected msf_in_export: speculative constant type checker: 
                                  the arguments { p } need to be MSFs, this is not allowed for export functions.
Failed as expected should_be_a_msf: speculative constant type checker: p should be an MSF
Failed as expected at_least_transient: speculative constant type checker: security annotation for p should be at least transient
Failed as expected unbound_level: speculative constant type checker: unbound security level hello
Failed as expected inconsistent_constraint: speculative constant type checker: cannot add constraint secret <= public, it is inconsistent
Failed as expected bad_modmsf: speculative constant type checker: annotation modmsf should be nomodmsf
Failed as expected bad_nomodmsf: speculative constant type checker: annotation nomodmsf should be modmsf
Failed as expected call_bad_nomodmsf: speculative constant type checker: annotation nomodmsf should be modmsf
Failed as expected call_bad_nomodmsf2: speculative constant type checker: annotation nomodmsf should be modmsf
Failed as expected call_bad_nomodmsf3: speculative constant type checker: annotation nomodmsf should be modmsf
Failed as expected call_modmsf_destroys: speculative constant type checker: 
                                         this function call destroys MSFs and { msf } are required.
                                         Trace:
                                           the function modmsf_destroys destroys MSFs at "error_messages.jazz", line 140 (2) to line 142 (3)
Failed as expected ret_high: speculative constant type checker: return type for p is #secret it should be less than #public
Failed as expected ret_transient: speculative constant type checker: return type for p is #transient it should be less than #public
Failed as expected ret_msf: speculative constant type checker: return annotation for msf should be msf
Failed as expected public_arg: speculative constant type checker: security annotation for x should be at least transient
Failed as expected unknown function: speculative constant type checker: unknown function unknown function
Failed as expected need_declassify: speculative constant type checker: return type for x is #secret it should be less than #public
